### ISSUE TITLE
Implement threshold per solution

### ISSUE THREAD

**Issue by jaubin** [tonerdo/coverlet#598]
Hi,

As of now we have the coverage threshold which works only on a per-test-project basis. Now let's imagine a solution which has several test projects, some are under the threshold, some are above the threshold. Globally the solution is above the threshold and this is what we want to check, while not caring about the results of individual test projects.

A possible workaround would be to put all the tests in one project but this is not satisfactory as a way to retrieve tests related to another module.

Would it be possible to set up a threshold mechanism at the solution level ? I guess it would be possible by delaying the threshold computation when we have a /P:MergeWith parameter (and another parameter as well).

Thanks a lot

**Comment by MarcoRossignoli** [comment]
Hi Julien, at the moment there is no way to run coverage solution wide.
You could try to manually parse output coverage file result and verify globally with a script.

dup of https://github.com/tonerdo/coverlet/issues/357
Discussion on possible solution https://github.com/tonerdo/coverlet/issues/542

Feel free to close if answered!

**Comment by jaubin** [comment]
OK thank you anyway.

I hope sooner or later we can reach a solution to do so. I close the ticket as it is some kind of duplicate.

Other possible solution : 
1/ do a mergeWith
2/ with command line compute the overall coverage by parsing the output file and make it above the threshold. Would this one possible ?

**Comment by jaubin** [comment]
I've seen and experienced this. The only missing feature would be coverlet console eating a report and breaking if the overall coverage does not reach the threshold.

**Comment by jaubin** [comment]
Another potential solution would be to give the coverlet command line a list of report file so that it does the computation, would it be possible ?

**Comment by MarcoRossignoli** [comment]
> Another potential solution would be to give the coverlet command line a list of report file so that it does the computation, would it be possible ?

Mmm I don't like partial solution...I think that we should complete solution wide coverage with this in mind.
If you need to calculate by yourself you can copy our internal code https://github.com/tonerdo/coverlet/blob/master/src/coverlet.core/CoverageResult.cs#L113

**Comment by jaubin** [comment]
Okay could do it using a for loop in a bash script, by extracting from the XML file the coverage percentage. However as of now I don't close the ticket because I guess we should get some more handy way to merge all the reports in a file and put a threshold which is only enforced by the last element.

It would be something like :
dotnet test <coverlet> /P:GlobalThreshold=80 

In that case it generates a global json alongside the other coverage results, and computes the final threshold from this json once the last test project has been passed. (Dunno if the latter is possible, but it would be great).

**Comment by MarcoRossignoli** [comment]
> In that case it generates a global json alongside the other coverage results, and computes the final threshold from this json once the last test project has been passed. (Dunno if the latter is possible, but it would be great).

It's not possible at the moment because tests could run in parallel if you run on sln so we don't have a clear way at the moment to put in relation different running test every test span custom process coverlet will be injected and produce "new file" but it doesn't know nothing about others.
TL;DR; This behaviour should be addressed if/when we'll support solution wide coverage. Again today coverlet is used through 3 drivers, msbuild(injected by dotnet test or msbuild) , .net tool and with collectors. We don't have control of all these engines, so at the moment it's not clear how support solution coverage "synch" with all drivers.

**Comment by mrpantsuit** [comment]
> Hi Julien, at the moment there is no way to run coverage solution wide.
> You could try to manually parse output coverage file result and verify globally with a script.
> 
> dup of #357
> Discussion on possible solution #542
> 
> Feel free to close if answered!

Sorry, is this really true?  Following the answer to this https://stackoverflow.com/questions/53255065/dotnet-unit-test-with-coverlet-how-to-get-coverage-for-entire-solution-and-not I was able to run coverage for my entire solution.  I'm running into the same problem as @jaubin, though, in that the Threshold is being applied to each project as it is run.  I want to be able to apply the threshold to the result entire solution, e.g., EntireSolutionThreshold=80.

**Comment by MarcoRossignoli** [comment]
Mmm...I'm not sure to understand...at the moment with msbuild driver and 
`dotnet test /p:CollectCoverage=true MergeWith.sln` (you can test using sample https://github.com/tonerdo/coverlet/tree/master/Documentation/Examples/MSBuild/MergeWith)

Msbuild create and run different instance of tasks that are overlapping, we can check this using MSBuild Binary and Structured Log Viewer.

![image](https://user-images.githubusercontent.com/7894084/68409944-4602f380-0188-11ea-9e60-63b2fcbb03d9.png)


And this is the timeline of events

![image](https://user-images.githubusercontent.com/7894084/68410061-67fc7600-0188-11ea-96be-ebbe5b26dd1e.png)


So every run write in same file https://github.com/tonerdo/coverlet/blob/545b8fc912562b8f72e3a392367d27d57e39044a/src/coverlet.msbuild.tasks/CoverageResultTask.cs#L128
And when we calculate the coverage we read same file and merge with "current" result https://github.com/tonerdo/coverlet/blob/4f3f25fda9badcb50aac83e182c79bbc4c7ff8df/src/coverlet.core/Coverage.cs#L281

I mean it should work with `dotnet test /p:CollectCoverage=true MergeWith.sln /m:1` where with m:1 we serialize msbuild.


So I don't understand how it can work, but maybe I'm missing something...tomorrow I'll take another look to understand better.

**Comment by mrpantsuit** [comment]
It seemed like it ran my test projects sequentially, but perhaps not.  If indeed it's running the test projects concurrently then yes, that would be problematic for aggregating test coverage data.

**Comment by MarcoRossignoli** [comment]
@mrpantsuit found also an old issue with this technique https://github.com/tonerdo/coverlet/issues/357#issuecomment-490404820

**Comment by jaubin** [comment]
Hi,
Yup there's an issue : if the last project does not actually contain tests global coverage is not computed at all.
    Le jeudi 7 novembre 2019 à 22:13:25 UTC+1, Marco Rossignoli <notifications@github.com> a écrit :  
 
 
@mrpantsuit found also an old issue with this technique #357 (comment)

**Comment by MarcoRossignoli** [comment]
Close for dup
